docs/source/daml/intro/5_Restrictions.rst

```diff
@@ -0,0 +1,103 @@
+
+.. literalinclude:: daml/Intro_5_Restrictions.daml
+  :language: daml
+  :start-after: -- REDEEM_CHOICE_BEGIN
+  :end-before: -- REDEEM_CHOICE_END
+
+.. literalinclude:: daml/Intro_5_Restrictions.daml
+  :language: daml
+  :start-after: -- REDEEM_TEST_BEGIN
+  :end-before: -- REDEEM_TEST_END
+
+There are quite a few new time-related functions here so it's time to talk about time on DAML ledgers. There's also quite a lot going on inside the ``do`` block of the ``Redeem`` choice, with several uses of the ``<-`` operator, which deserves a proper explanation at this point.
+
+Time on DAML ledgers
+--------------------
+
+Each transaction on a DAML ledger has two timestamps called the *ledger effective time (LET)* and the *record time (RT)*. The ledger effective time is set by the submitter of a transaction, the record time is set by the consensus protocoll. Each DAML ledger has a policy on the allowed difference between LET and RT called the *skew*. The submitter has to take a good guess at what the record time will be. If it's too far off, the transaction will be rejected.
+
+``getTime`` is an action that gets the LET from the ledger. In the above example, that time is taken apart into day of week and hour of day using standard library functions from ``DA.Date`` and ``DA.Time``. The hour of the day is checked to be in the range from 8 to 18. Suppose now that the ledger had a skew of 10 seconds, but a submission took less than 4 seconds to commit. At 18:00:05, Alice could submit a transaction to redeem an Iou with a LET of 17:59:59. It would be a valid trasnaction and be committed successfully as ``hrs == 17`` and ``LET - RT < 10 seconds``. Time therefore has to be considered slightly fuzzy in DAML, with the fuzzyness depending on the skew parameter.
+
+In scenarios, record and ledger effective time are always equal and they can be set using the functions ``passToDate`` and ``pass``. ``passToDate`` takes a date sets the time to midnight (UTC) of that date. ``pass`` takes a ``Reltime``, a relative time and moves the ledger by that much.
+
+On a distributed DAML ledger, there are no guarantees that ledger effective time or relative time are strictly increasing. The only guarantee is that ledger effective time is increasing with causality. If a transaction ``TX2`` depends on a transaction ``TX1``, then the ledger enforces that the LET of ``TX2`` is greater than or equal to that of ``TX1``:
+
+.. literalinclude:: daml/Intro_5_Restrictions.daml
+  :language: daml
+  :start-after: -- CAUSALITY_TEST_BEGIN
+  :end-before: -- CAUSALITY_TEST_END
+
+Actions and ``do`` blocks
+-------------------------
+
+You have come across ``do`` blocks and ``<-`` notations in two contexts by now: ``Scenario`` and ``Update``. Both of these are examples of an ``Action``, also called a *Monad* in functional programming.
```

Comment:
Could you add a WIIFM sentence here? Perhaps something like "Understanding what an ``Action`` is will help you do..."

Reply:
```suggestion
You have come across ``do`` blocks and ``<-`` notations in two contexts by now: ``Scenario`` and ``Update``. Both of these are examples of an ``Action``, also called a *Monad* in functional programming. ``Actions`` can be constructed in a convenient way using ``do`` notation. Understanding ``Actions`` and ``do`` blocks is therefore crucial to being able to construct correct contract models and test.
```